File: src/strarr/strarr.nim
import times

type PGoStrSlice = pointer

proc genStrSlice(n:int):PGoStrSlice {.dynlib: "./strarr.so", importc: "genStrSlice".}
proc `[]`(p: PGoStrSlice, i:int):cstring {.dynlib: "./strarr.so", importc: "getGoStrSliceVal".}
proc len(p: PGoStrSlice):int {.dynlib: "./strarr.so", importc: "getGoStrSliceLen".}

proc `@`(p: PGoStrSlice):seq[cstring] =
  let sliceLen = p.len()
  var s = newSeq[cstring](sliceLen)
  for i in 0..<sliceLen:
    s[i] = p[i]
  return s

when isMainModule:
  const length = 10000
  let start = cpuTime()
  let strSlice = @( genStrSlice(length) )
  echo cpuTime() - start
  echo strSlice.len
  echo strSlice[0]
  echo strSlice[length-1]

File: src/strarr/strarr.go
package main

import (
	"C"
	"unsafe"
)

type GoStrSlice = []string
type PGoStrSlice = uintptr

var boxInstances map[uintptr]*GoStrSlice

//export genStrSlice
func genStrSlice() PGoStrSlice {
	s := GoStrSlice{"あい", "うえ", "おか"}

	p := PGoStrSlice(unsafe.Pointer(&s))
	if boxInstances == nil {
		boxInstances = make(map[PGoStrSlice]*GoStrSlice)
	}
	boxInstances[p] = &s
	return p
}

//export getGoStrSliceVal
func getGoStrSliceVal(p PGoStrSlice, i int) *C.char {
	slice := (*GoStrSlice)(unsafe.Pointer(p))
	return C.CString((*slice)[i])
}

//export getGoStrSliceLen
func getGoStrSliceLen(p PGoStrSlice) int {
	s := (*GoStrSlice)(unsafe.Pointer(p))
	return len(*s)
}

func main() {}
